sdks/v4-sdk/src/utils/v4Planner.ts

```diff
@@ -0,0 +1,131 @@
+
+const POSITION_CONFIG_STRUCT = '(' + POOL_KEY_STRUCT + ' poolKey,int24 tickLower,int24 tickUpper)'
+
+const SWAP_EXACT_IN_SINGLE_STRUCT =
+  '(' +
+  POOL_KEY_STRUCT +
+  ' poolKey,bool zeroForOne,uint128 amountIn,uint128 amountOutMinimum,uint160 sqrtPriceLimitX96,bytes hookData)'
+
+const SWAP_EXACT_IN_STRUCT =
+  '(address currencyIn,' + PATH_KEY_STRUCT + '[] path,uint128 amountIn,uint128 amountOutMinimum)'
+
+const SWAP_EXACT_OUT_SINGLE_STRUCT =
+  '(' +
+  POOL_KEY_STRUCT +
+  ' poolKey,bool zeroForOne,uint128 amountOut,uint128 amountInMaximum,uint160 sqrtPriceLimitX96,bytes hookData)'
+
+const SWAP_EXACT_OUT_STRUCT =
+  '(address currencyOut,' + PATH_KEY_STRUCT + '[] path,uint128 amountOut,uint128 amountInMaximum)'
+
+const ABI_DEFINITION: { [key in Actions]: string[] } = {
+  // Liquidity commands
+  [Actions.INCREASE_LIQUIDITY]: ['uint256', POSITION_CONFIG_STRUCT, 'uint256', 'uint128', 'uint128', 'bytes'],
+  [Actions.DECREASE_LIQUIDITY]: ['uint256', POSITION_CONFIG_STRUCT, 'uint256', 'uint128', 'uint128', 'bytes'],
+  [Actions.MINT_POSITION]: [POSITION_CONFIG_STRUCT, 'uint256', 'uint128', 'uint128', 'address', 'bytes'],
+  [Actions.BURN_POSITION]: ['uint256', POSITION_CONFIG_STRUCT, 'uint128', 'uint128', 'bytes'],
+
+  // Swapping commands
+  [Actions.SWAP_EXACT_IN_SINGLE]: [SWAP_EXACT_IN_SINGLE_STRUCT],
+  [Actions.SWAP_EXACT_IN]: [SWAP_EXACT_IN_STRUCT],
+  [Actions.SWAP_EXACT_OUT_SINGLE]: [SWAP_EXACT_OUT_SINGLE_STRUCT],
+  [Actions.SWAP_EXACT_OUT]: [SWAP_EXACT_OUT_STRUCT],
+
+  // Payments commands
+  [Actions.SETTLE]: ['address', 'uint256', 'bool'],
+  [Actions.SETTLE_ALL]: ['address', 'uint256'],
+  [Actions.TAKE]: ['address', 'address', 'uint256'],
+  [Actions.TAKE_ALL]: ['address', 'uint256'],
+  [Actions.TAKE_PORTION]: ['address', 'address', 'uint256'],
+  [Actions.SETTLE_TAKE_PAIR]: ['address', 'address'],
+  [Actions.CLOSE_CURRENCY]: ['address'],
+  [Actions.SWEEP]: ['address', 'address'],
+}
+
+export class V4Planner {
+  actions: string
+  params: string[]
+
+  constructor() {
+    this.actions = '0x'
+    this.params = []
+  }
+
+  addAction(type: Actions, parameters: any[]): void {
+    let command = createAction(type, parameters)
+    this.params.push(command.encodedInput)
+    this.actions = this.actions.concat(command.action.toString(16).padStart(2, '0'))
+  }
+
+  addTrade(trade: Trade<Currency, Currency, TradeType>, slippageTolerance: Percent): void {
+    const actionType = trade.tradeType === TradeType.EXACT_INPUT ? Actions.SWAP_EXACT_IN : Actions.SWAP_EXACT_OUT
+    const inputCurrency = trade.inputAmount.currency
+
+    for (let swap of trade.swaps) {
+      this.addAction(actionType, [
+        {
+          inputCurrency.isNative ? ADDRESS_ZERO : inputCurrency.wrapped.address,
+          path: encodeRouteToPath(swap.route),
+          amountIn: swap.inputAmount,
+          amountOutMinimum: /* need to do slippage checks for each individual route??? */,
```

Comment:
@zhongeric this is what kinda stopped me in my tracks for a second

Reply:
Ah yes so this is a bit tricky - looking at the UR sdk code, for every `swap` in the trade, it generates a new Trade from that route, and calculates the amountOutMinimum for each subTrade within the larger Trade.

I don't think we need to do any of this swap/liquidity specific logic here in the planner. The classes can be in charge of adding the correct Actions in the correct order IMO

Reply:
Heyy, not exactly sure what you mean by your second statement here. Since everything that needs to happen in a V4 Trade is encapsulated within an Action that's submitted to a general V4Command, I wanted to abstract that away completely into the V4 planner for separation of concerns etc etc. so I think I will include a slippage check arg here. 

let me know if I misunderstood what you're saying, if you have an alternative suggestion!